### ISSUE TITLE
Stack Overflow in `ReceiveMessage`/`SendMessage`

### ISSUE THREAD

**Issue by iwillspeak** [dotnet/orleans#6559]
We have a production cluster with ~20 instances. When under load we get the occasional patches of silo crashes due to "Stack overflow.". 

From the stack trace it appears that the runtime is forwarding a message that it thinks should be handled by another silo, but the current silo ends up receiving it again instead.

We're running Orleans 3.1.6 on .NET Core 3.1. Host is an Ubuntu 18 instance.

```
(lldb) Process 7962 stopped
* thread #40: tid = 8326, 0x00007fa00661b211 libcoreclr.so`___lldb_unnamed_symbol8948$$libcoreclr.so + 17, name = 'dotnet', stop reason = signal SIGSEGV: address access protected (fault address: 0x7f99f2ffeffc)
    frame #0: 0x00007fa00661b211 libcoreclr.so`___lldb_unnamed_symbol8948$$libcoreclr.so + 17
libcoreclr.so`___lldb_unnamed_symbol8948$$libcoreclr.so:
->  0x7fa00661b211 <+17>: movl   %r9d, -0x34(%rbp)
    0x7fa00661b215 <+21>: movl   %r8d, -0x30(%rbp)
    0x7fa00661b219 <+25>: movq   %rcx, -0x58(%rbp)
    0x7fa00661b21d <+29>: movq   %rdx, -0x48(%rbp)
(lldb) clrstack
OS Thread Id: 0x2086 (40)
        Child SP               IP Call Site
00007F99F2FFF1E8 00007fa00661b211 [HelperMethodFrame: 00007f99f2fff1e8] 
00007F99F2FFF330 00007F9F958024C5 System.String.Ctor(System.ReadOnlySpan`1<Char>) [/_/src/System.Private.CoreLib/shared/System/String.cs @ 352]
00007F99F2FFF360 00007F9F96DE6DA6 System.Text.StringBuilder.AppendFormatHelper(System.IFormatProvider, System.String, System.ParamsArray) [/_/src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs @ 1553]
00007F99F2FFF430 00007F9F96E2B5C1 Orleans.Runtime.UniqueKey.ToHexString()
00007F99F2FFF480 00007F9F96E409E0 Orleans.Runtime.GrainId.ToStringImpl(Boolean)
00007F99F2FFF510 00007F9F96DE6E07 System.Text.StringBuilder.AppendFormatHelper(System.IFormatProvider, System.String, System.ParamsArray) [/_/src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs @ 1553]
00007F99F2FFF5E0 00007F9F96DE88CD System.String.FormatHelper(System.IFormatProvider, System.String, System.ParamsArray) [/_/src/System.Private.CoreLib/shared/System/String.Manipulation.cs @ 525]
00007F99F2FFF620 00007F9F96E3D97D System.String.Format(System.String, System.Object[]) [/_/src/System.Private.CoreLib/shared/System/String.Manipulation.cs @ 486]
00007F99F2FFF680 00007F9F96E51774 Orleans.Runtime.Message.ToString()
00007F99F2FFF6F0 00007F9F96DE6E07 System.Text.StringBuilder.AppendFormatHelper(System.IFormatProvider, System.String, System.ParamsArray) [/_/src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs @ 1553]
00007F99F2FFF7C0 00007F9F96DE88CD System.String.FormatHelper(System.IFormatProvider, System.String, System.ParamsArray) [/_/src/System.Private.CoreLib/shared/System/String.Manipulation.cs @ 525]
00007F99F2FFF800 00007F9F96E3B9AF Orleans.Runtime.Messaging.IncomingMessageAgent.ReceiveMessage(Orleans.Runtime.Message)
00007F99F2FFF880 00007F9F96E3B489 Orleans.Runtime.Messaging.OutboundMessageQueue.SendMessage(Orleans.Runtime.Message)
00007F99F2FFF960 00007F9F96E3B9E6 Orleans.Runtime.Messaging.IncomingMessageAgent.ReceiveMessage(Orleans.Runtime.Message)

------------------------- 8< ------------------------- snip ------------------------- >8 ------------------------- 

00007F99F37FD040 00007F9F96E3B489 Orleans.Runtime.Messaging.OutboundMessageQueue.SendMessage(Orleans.Runtime.Message)
00007F99F37FD120 00007F9F96E3B9E6 Orleans.Runtime.Messaging.IncomingMessageAgent.ReceiveMessage(Orleans.Runtime.Message)
00007F99F37FD1A0 00007F9F96E3B489 Orleans.Runtime.Messaging.OutboundMessageQueue.SendMessage(Orleans.Runtime.Message)
00007F99F37FD280 00007F9F96E3B9E6 Orleans.Runtime.Messaging.IncomingMessageAgent.ReceiveMessage(Orleans.Runtime.Message)
00007F99F37FD300 00007F9F96E3B489 Orleans.Runtime.Messaging.OutboundMessageQueue.SendMessage(Orleans.Runtime.Message)
00007F99F37FD3E0 00007F9F972EEB68 Orleans.Runtime.Dispatcher.ResendMessageImpl(Orleans.Runtime.Message, Orleans.Runtime.ActivationAddress)
00007F99F37FD430 00007F9F972ED450 Orleans.Runtime.Dispatcher.TryForwardMessage(Orleans.Runtime.Message, Orleans.Runtime.ActivationAddress)
00007F99F37FD460 00007F9F972ED0DC Orleans.Runtime.Dispatcher.TryForwardRequest(Orleans.Runtime.Message, Orleans.Runtime.ActivationAddress, Orleans.Runtime.ActivationAddress, System.String, System.Exception)
00007F99F37FD500 00007F9F97348486 Orleans.Runtime.Dispatcher+<>c__DisplayClass30_0.<ProcessRequestsToInvalidActivation>b__0()
00007F99F37FD530 00007F9F96E335CD Orleans.Runtime.Scheduler.TaskSchedulerUtils.RunWorkItemTask(Orleans.Runtime.Scheduler.IWorkItem)
00007F99F37FD550 00007F9F9580658D System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object) [/_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs @ 201]
00007F99F37FD5A0 00007F9F96E03AFD System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread) [/_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs @ 2421]
00007F99F37FD620 00007F9F96E0A4DB System.Threading.Tasks.Task.ExecuteEntry() [/_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs @ 2303]
00007F99F37FD640 00007F9F96E0DC49 Orleans.Runtime.Scheduler.ActivationTaskScheduler.RunTask(System.Threading.Tasks.Task)
00007F99F37FD690 00007F9F96CF4270 Orleans.Runtime.Scheduler.WorkItemGroup.Execute()
00007F99F37FD750 00007F9F96E502F3 Orleans.Runtime.Scheduler.OrleansSchedulerAsynchAgent.Process(Orleans.Runtime.Scheduler.IWorkItem)
00007F99F37FD770 00007F9F96CF0499 Orleans.Threading.FiltersApplicant`1[[System.__Canon, System.Private.CoreLib]].Apply(System.__Canon)
00007F99F37FD7B0 00007F9F96E4FD83 Orleans.Threading.ExecutionContext.ExecuteWithFilters(Orleans.Threading.WorkItem)
00007F99F37FD7D0 00007F9F96CED1C3 Orleans.Threading.ThreadPoolExecutor.ProcessWorkItems(Orleans.Threading.ExecutionContext)
00007F99F37FD830 00007F9F96CECC37 Orleans.Threading.ThreadPoolExecutor+<>c__DisplayClass13_0.<RunWorker>b__0(System.Object)
00007F99F37FD850 00007F9F96CEC6C6 Orleans.Threading.ThreadPoolThread+<>c__DisplayClass8_0.<QueueWorkItem>b__0()
00007F99F37FD890 00007F9F8CC554B2 System.Threading.ThreadHelper.ThreadStart_Context(System.Object) [/_/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs @ 44]
00007F99F37FD8B0 00007F9F9580658D System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object) [/_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs @ 201]
00007F99F37FD900 00007F9F8CC555CE System.Threading.ThreadHelper.ThreadStart() [/_/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs @ 93]
00007F99F37FDC60 00007fa00655a5ef [GCFrame: 00007f99f37fdc60] 
00007F99F37FDD30 00007fa00655a5ef [DebuggerU2MCatchHandlerFrame: 00007f99f37fdd30] 
(lldb) 
```
 
[Full stack trace](https://github.com/dotnet/orleans/files/4697084/Orleans.Stack.Overflow.txt)

Any idea what could be going wrong here.

**Comment by sergeybykov** [comment]
This was fixed in 3.1.7 via #6525.

**Comment by iwillspeak** [comment]
Awesome thanks. I searched but couldn’t find a matching issue.
